### PR TITLE
WIP: Updated code standards and expectations

### PR DESCRIPTION
**Pull request by tophmatthews**
This just captures the standards document I shared earlier. It'll need a bit of work, and decisions are needed to add what where, but it's something to start talking over. @novasr @permcody @bwspenc @dschwen

### CODE REVIEW

File: modules/doc/content/framework_development/code_standards.md [hunk 1]
```diff
@@ -289,6 +291,47 @@ protected:
...
+## Best practices
+
+-	All input parameters should always use range checked parameters where applicable
+-	Input parameters should always use MooseEnum instead of unsigned int
+-	mooseAssert should always be used to clearly state coding assumptions and to protect dividing by zero
+-	mooseException should be used in cases where the conditions typically used for a mooseAssert are likely to occur (e.g. inner Newton iteration that produces a negative value).
+-	mooseException should be used in favor of mooseError when checking run time conditions
+-	paramError/paramWarning should be used in favor of mooseError/mooseWarning when checking input file conditions
+-	Variable names shall be used that describe the parameter in favor of small names (e.g. activation_energy instead of Q)
+-	Utility::pow<n> shall be used in place of std::pow(x,n) if the power is constant
```
Comment (bwspenc) [inline modules/doc/content/framework_development/code_standards.md]: if the power is a constant integer.  If the power is non-constant integer, use MathUtils::pow(x,n).

File: modules/doc/content/framework_development/code_standards.md [hunk 2]
```diff
@@ -289,6 +291,47 @@ protected:
...
+
+
+## Code testing
+
+Goal: 1) ensure what is in the documentation is what is in the code, and 2) ensure the code works across a wide swath of parameters
+
+-	MOOSE SQA standards shall always be followed
```
Comment (bwspenc) [inline modules/doc/content/framework_development/code_standards.md]: I think this needs to be more specific.

File: modules/doc/content/framework_development/code_standards.md [hunk 3]
```diff
@@ -289,6 +291,47 @@ protected:
...
+-	Default values shall be overridden in the test file whenever possible.
```
Comment (bwspenc) [inline modules/doc/content/framework_development/code_standards.md]: Not sure what you mean by this. Do you mean in the .i file or the 'tests' file.  Also, why is that important?

File: modules/doc/content/framework_development/code_standards.md [hunk 4]
```diff
@@ -289,6 +291,47 @@ protected:
...
+-	Input parameter values shall be tested over the full range of applicability, e.g. across all concentrations of constituents and/or a wide variety of temperatures.
+-	Hand calculations, either as simple comments in the input file, added as functions in the input file, or in an adjoining python/excel file should always be included where possible. The equations used to get the hand calculations shall always be included, not just the final answers.
+-	All input files should be as simple as possible, with any unnecessary physics, input file fluff, and variables removed (e.g. `solve = false` for tests on materials only)
```
Comment (bwspenc) [inline modules/doc/content/framework_development/code_standards.md]: I definitely agree that making tests that are as close to unit tests as possible is good, but it's probably to say that 'all input files' should do that. We also need integration tests (as you say below)

**Comment by bwspenc** [comment]
Not stale. This just needs a small amount of tweaking.

File: modules/doc/content/framework_development/code_standards.md [hunk 5]
```diff
@@ -289,6 +291,47 @@ protected:
...
+- Every destructor must be virtual.
+- All function definitions should be in *.C files.
+    - The only exceptions are for inline functions for speed and templates.
+- Thou shall not commit accidental insertion in a std::map by using brackets in a RHS operator unless he/she can prove it can't fail.
+- Thou shall use range-based loops or MooseIndex() based loops for iteration
+
+## Best practices
+
+-	All input parameters should always use range checked parameters where applicable
+-	Input parameters should always use MooseEnum instead of unsigned int
```
Comment (dschwen) [inline modules/doc/content/framework_development/code_standards.md]: Needs to be more specific. Maybe "input parameters accepting a finite set of permitted options" (then again the number of unsigned int values is finite as well :-D )

File: modules/doc/content/framework_development/code_standards.md [hunk 6]
```diff
@@ -289,6 +291,47 @@ protected:
...
+- Use references instead of pointers whenever possible
+    - i.e., this object lives for a shorter period of time than the object it needs to refer to does
+- Methods should return pointers to objects if returned objects are stored as pointers and references if returned objects are stored as references
+- When creating a new class:
+    - Include dependent header files in the *.C file whenever possible (i.e. the header uses only references or pointers in it's various declarations) and use forward declarations in the header file as needed
+    - One exception is when doing so would require end users to include multiple files to complete definitions of child objects (Errors typically show up as "incomplete class" or something similar during compilation)
+- Avoid using a global variable when possible.
+- Every destructor must be virtual.
+- All function definitions should be in *.C files.
+    - The only exceptions are for inline functions for speed and templates.
```
Comment (dschwen) [inline modules/doc/content/framework_development/code_standards.md]: "inline functions for speed" is misleading. This is not what inline does these days (it is what it used to do in the '90s :-D)

File: modules/doc/content/framework_development/code_standards.md [hunk 7]
```diff
@@ -289,6 +291,47 @@ protected:
...
+- Thou shall not commit accidental insertion in a std::map by using brackets in a RHS operator unless he/she can prove it can't fail.
+- Thou shall use range-based loops or MooseIndex() based loops for iteration
```
Comment (dschwen) [inline modules/doc/content/framework_development/code_standards.md]: Yeah, cute, but not appropriate language for a document like this

File: modules/doc/content/framework_development/code_standards.md [hunk 8]
```diff
@@ -289,6 +291,47 @@ protected:
...
+
+## Best practices
+
+-	All input parameters should always use range checked parameters where applicable
+-	Input parameters should always use MooseEnum instead of unsigned int
+-	mooseAssert should always be used to clearly state coding assumptions and to protect dividing by zero
+-	mooseException should be used in cases where the conditions typically used for a mooseAssert are likely to occur (e.g. inner Newton iteration that produces a negative value).
+-	mooseException should be used in favor of mooseError when checking run time conditions
```
Comment (dschwen) [inline modules/doc/content/framework_development/code_standards.md]: No. It should be used where aborting a solve could recover from the error.

File: modules/doc/content/framework_development/code_standards.md [hunk 9]
```diff
@@ -289,6 +291,47 @@ protected:
...
+
+
+## Code testing
+
+Goal: 1) ensure what is in the documentation is what is in the code, and 2) ensure the code works across a wide swath of parameters
+
+-	MOOSE SQA standards shall always be followed
+-	Default values shall be overridden in the test file whenever possible.
+-	Input parameter values shall be tested over the full range of applicability, e.g. across all concentrations of constituents and/or a wide variety of temperatures.
+-	Hand calculations, either as simple comments in the input file, added as functions in the input file, or in an adjoining python/excel file should always be included where possible. The equations used to get the hand calculations shall always be included, not just the final answers.
```
Comment (dschwen) [inline modules/doc/content/framework_development/code_standards.md]: I'd rather leave excel out of this document. Python covers us fully and can be run automatically by the test harness.

File: modules/doc/content/framework_development/code_standards.md [hunk 10]
```diff
@@ -289,6 +291,47 @@ protected:
...
+-	All input files should be as simple as possible, with any unnecessary physics, input file fluff, and variables removed (e.g. `solve = false` for tests on materials only)
+-	All tests should be tested using CSVDiff unless Exodiff is required to capture the change during the simulation caused by the new code.
```
Comment (dschwen) [inline modules/doc/content/framework_development/code_standards.md]: Hard to say that in advance. We mainly want to protect against breakage from _future_ modifications. I'd strike this point or weaken it considerably.

File: modules/doc/content/framework_development/code_standards.md [hunk 11]
```diff
@@ -306,17 +349,9 @@ class MyClass:
...
-- All function definitions should be in *.C files.
-    - The only exceptions are for inline functions for speed and templates.
-- Thou shall not commit accidental insertion in a std::map by using brackets in a RHS operator unless he/she can prove it can't fail.
-- Thou shall use range-based loops or MooseIndex() based loops for iteration
\ No newline at end of file
+# Documentation:
+  Goal: Present the code in words and documents that give background to the choices made in the code
+  
+- Documentation shall include any equations, definitions of the variables, appropriate referencing of the equations to external publications, description of the code, and use case for the code.
+- Best practices used for journal publications should be adopted
```
Comment (dschwen) [inline modules/doc/content/framework_development/code_standards.md]: at least provide a link to such a best practices document

**Review by dschwen — CHANGES_REQUESTED** [review]

**Comment by tophmatthews** [comment]
All great suggestions @dschwen. My original intent was to get up a document we discussed on a call with @bwspenc, @permcody, and @novasr, and have everyone contribute. That document was targeted for BISON originally, but @permcody identified most of the bullets are relevant to all MOOSE apps. How about everyone just adding/removing the things you think relevant directly? I think you all can do that to my branch?

**Comment by permcody** [comment]
I don't want to lose this. @tophmatthews - do you want me to take ownership of this? I can push changes that I want (and incorporate Ben and Daniel's) ideas as well. I think most of what you have is fine. I might change a few things, but I do want most of this at the MOOSE level.

**Comment by tophmatthews** [comment]
> do you want me to take ownership of this

That would be great @permcody, it's probably better this says what you'd like to see. Thanks!

**Pull request closed without merging**